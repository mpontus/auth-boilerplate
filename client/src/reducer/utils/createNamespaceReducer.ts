import { Action, Reducer } from "redux";

type Namespace<K extends string, V> = Partial<Record<K, V>>;

type HigherOrderReducer<S, A extends Action<any>> = (
  action: A
) => (state: S, action: A) => S;

export const createNamespaceReducer = <K extends string, V, A extends Action>(
  fn: HigherOrderReducer<Namespace<K, V>, A>
): Reducer<Namespace<K, V>, A> => (state = {}, action) =>
  fn(action)(state, action);
