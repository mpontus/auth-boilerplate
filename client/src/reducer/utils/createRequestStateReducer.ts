import { Action, Reducer } from "redux";

interface RequestState<ErrorType> {
  loading: boolean;
  success: boolean;
  error?: ErrorType;
}

const initialState = {
  loading: false,
  success: false
};

/**
 * Reusable reducer constructor for managing request state
 */
export const createRequestStateReducer = <
  ActionType extends Action<any>,
  ErrorType extends Error
>(
  reducer: (
    state: RequestState<ErrorType>,
    action: ActionType
  ) => RequestState<ErrorType>
): Reducer<RequestState<ErrorType>, ActionType> => (
  state = initialState,
  action
) => reducer(state, action);
