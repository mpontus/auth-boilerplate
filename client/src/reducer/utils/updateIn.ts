import { Action, Reducer } from "redux";

/**
 * Return a function which updates given key in an object, using the
 * provided updater function.
 */
export const updateIn = <
  K extends keyof R,
  A extends Action,
  V,
  R = Record<K, V>
>(
  key: K,
  updater: Reducer<R[K], A>
) => (state: R, action: A): R =>
  Object.assign({}, state, { [key]: updater(state[key], action) });
