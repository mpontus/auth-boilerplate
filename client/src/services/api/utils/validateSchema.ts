import * as t from "io-ts";
import { PathReporter } from "io-ts/lib/PathReporter";

export const validateSchema = <TSchema extends t.Type<any>>(
  schema: TSchema,
  data: any
): t.TypeOf<TSchema> => {
  const result = schema.decode(data);

  if (result.isLeft()) {
    throw new Error(PathReporter.report(result).join("\n"));
  }

  return result.value;
};
