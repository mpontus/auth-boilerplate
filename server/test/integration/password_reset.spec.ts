import request from 'supertest';
import { resetDb } from '../utils/resetDb';
import { initApp } from '../utils/initApp';

let expressApp: any, nestApp: any;

beforeAll(async () => {
  ({ expressApp, nestApp } = await initApp());
});

afterAll(() => nestApp.close());

beforeEach(() => resetDb());

describe('Password reset', () => {
  describe('when email and secret are correct', () => {
    const newPassword = 'foobarbaz';
    const seed = require('../seeds/password_recovery_request');

    beforeEach(() => seed.run());

    it('should return successful response', async () => {
      const response = await request(expressApp)
        .post('/auth/reset_password')
        .send({
          secret: seed.secret,
          email: seed.userEmail,
          password: newPassword,
        })
        .expect(202);

      expect(response.body).toMatchSnapshot();
    });
  });

  describe('when password is invalid', () => {
    const seed = require('../seeds/password_recovery_request');

    beforeEach(() => seed.run());

    it('should return failed response', async () => {
      const response = await request(expressApp)
        .post('/auth/reset_password')
        .send({
          secret: seed.secret,
          email: seed.userEmail,
          password: '',
        })
        .expect(400);

      expect(response.body).toMatchSnapshot();
    });
  });
});
