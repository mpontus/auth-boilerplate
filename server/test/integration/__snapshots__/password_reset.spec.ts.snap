// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Password reset when email and secret are correct should return successful response 1`] = `Object {}`;

exports[`Password reset when password is invalid should return failed response 1`] = `
Object {
  "error": "Bad Request",
  "message": Array [
    Object {
      "children": Array [],
      "constraints": Object {
        "minLength": "password must be longer than or equal to 6 characters",
      },
      "property": "password",
      "target": Object {
        "email": "Fannie.Brekke88@gmail.com",
        "password": "",
        "secret": "38169",
      },
      "value": "",
    },
  ],
  "statusCode": 400,
}
`;
